File: dotnet/EliteDangerousSdk/Planner/Materials.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace EliteDangerousSdk.Planner
{
    public class MaterialEntry
    {
        public string Name { get; set; } = "";
        public int? EdId { get; set; }
        public string Category { get; set; } = "";
        public int Grade { get; set; }
        public int Count { get; set; }
        public int MaxCapacity { get; set; } = 1000;
    }

    public class MaterialInventory
    {
        public List<MaterialEntry> Materials { get; set; } = new();
        public DateTime Timestamp { get; set; } = DateTime.UtcNow;
    }

    public class BlueprintRequirement
    {
        public string MaterialName { get; set; } = "";
        public int Quantity { get; set; }
        public int Grade { get; set; }
    }

    public static class MaterialCaps
    {
        public static readonly Dictionary<int, int> ByGrade = new()
        {
            [1] = 3000, [2] = 3000, [3] = 3000, [4] = 3000, [5] = 3000,
        };
    }

    public static class Materials
    {
        public static MaterialInventory CreateInventory()
        {
            return new MaterialInventory
            {
                Materials = new List<MaterialEntry>(),
                Timestamp = DateTime.UtcNow,
            };
        }

        public static MaterialInventory UpdateInventory(MaterialInventory inv, string materialName, int change)
        {
            var entry = inv.Materials.FirstOrDefault(m => m.Name == materialName);
            if (entry != null)
            {
                entry.Count = Math.Max(0, Math.Min(entry.MaxCapacity, entry.Count + change));
            }
            inv.Timestamp = DateTime.UtcNow;
            return inv;
        }

        public static (bool canCraft, List<(string name, int have, int need)> missing) CanCraftBlueprint(
            MaterialInventory inv, List<BlueprintRequirement> requirements)
        {
            var missing = new List<(string, int, int)>();
            bool canCraft = true;
            foreach (var req in requirements)
            {
                var entry = inv.Materials.FirstOrDefault(m => m.Name == req.MaterialName);
                var have = entry?.Count ?? 0;
                if (have < req.Quantity)
                {
                    canCraft = false;
                    missing.Add((req.MaterialName, have, req.Quantity));
                }
            }
            return (canCraft, missing);
        }
    }
}

File: python/elite_dangerous_sdk/planner/materials.py
from dataclasses import dataclass, field
from datetime import datetime
from typing import List, Optional, Tuple
from ..data import material


@dataclass
class MaterialEntry:
    name: str = ""
    ed_id: Optional[int] = None
    category: str = ""
    grade: int = 0
    count: int = 0
    max_capacity: int = 1000


@dataclass
class MaterialInventory:
    materials: List[MaterialEntry] = field(default_factory=list)
    timestamp: datetime = field(default_factory=datetime.utcnow)


@dataclass
class BlueprintRequirement:
    material_name: str = ""
    quantity: int = 0
    grade: int = 0


MATERIAL_CAPS: dict = {1: 3000, 2: 3000, 3: 3000, 4: 3000, 5: 3000}
MICRO_RESOURCE_CAPS: dict = {1: 3000, 2: 3000, 3: 3000, 4: 3000, 5: 3000}


def create_inventory() -> MaterialInventory:
    entries = []
    for m in material:
        entry = MaterialEntry(
            name=m.get("name", ""),
            ed_id=m.get("id"),
            category=m.get("category", ""),
            grade=m.get("grade", 0),
            max_capacity=3000,
        )
        entries.append(entry)
    return MaterialInventory(materials=entries)


def update_inventory(inv: MaterialInventory, material_name: str, change: int) -> MaterialInventory:
    for entry in inv.materials:
        if entry.name == material_name:
            entry.count = max(0, min(entry.max_capacity, entry.count + change))
            break
    inv.timestamp = datetime.utcnow()
    return inv


def can_craft_blueprint(inv: MaterialInventory, requirements: List[BlueprintRequirement]) -> Tuple[bool, List[Tuple[str, int, int]]]:
    missing = []
    can_craft = True
    for req in requirements:
        have = 0
        for entry in inv.materials:
            if entry.name == req.material_name:
                have = entry.count
                break
        if have < req.quantity:
            can_craft = False
            missing.append((req.material_name, have, req.quantity))
    return (can_craft, missing)
